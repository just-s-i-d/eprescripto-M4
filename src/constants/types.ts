export type InfoCardDetailsType =
  | {
      totalCount: number;
      title: string;
    }[]
  | undefined;

export type ErrorType = {
  message: string;
  type: string;
};
export type AppointmentDataType = {
  pId: string;
  email: string;
  lastVisited: string;
  avatar: string;
  gender: string;
  key: number;
  pName: string;
  timeSlot: string;
  contact: string;
  status: string;
  profilePic: string;
  referrer: string;
  date: string;
};

export type AppointmentsDataType = AppointmentDataType[] | undefined;

export type PrescriptionDataType = {
  key: React.Key;
  prescriptionId: string;
  name: string;
  duration: number;
  date: string;
  status: "Active" | "Completed" | "On Hold";
  notes: string;
};

export type PrescrtionsDataType = PrescriptionDataType[] | undefined;

export type ChartDataType = {
  labels: string[];
  data: number[];
  totalCount: number;
};

export type LineChartDataType = {
  labels: string[];
  data: number[];
};

export type LineChartApiResDataType = {
  seven: number[];
  ten: number[];
  thirty: number[];
};

export type ChartApiDataType = {
  pieChartData: ChartDataType;
  barChartData: ChartDataType;
  lineChartData: LineChartApiResDataType;
};

export type DoctorAppointmentsDataType = AppointmentDataType[];

export type ReviewDataType = {
  pId: string;
  pName: string;
  email: string;
  rating: 1 | 2 | 3 | 4 | 5;
  comment: string;
};
export type ReviewsDataType = ReviewDataType[];
