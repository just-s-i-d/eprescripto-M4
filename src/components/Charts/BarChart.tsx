import { Card, Skeleton } from "antd";
import {
  Chart as ChartJS,
  CategoryScale,
  LinearScale,
  BarElement,
  Title,
  Tooltip,
  Legend,
  scales,
} from "chart.js";
import { Bar } from "react-chartjs-2";
import type { ChartOptions } from "chart.js";

import ErrorBoundary from "@components/ErrorBoundary";
import CardTitle from "@components/ui/CardTitle";
import { ChartDataType } from "@constants/types";
import { useEffect, useState } from "react";
import UseStatesHook from "src/hooks/UseStatesHook";

ChartJS.register(
  CategoryScale,
  LinearScale,
  BarElement,
  Title,
  Tooltip,
  Legend,
  scales,
);

type PieChartPropsType = {
  chartData: ChartDataType | undefined;
  title: string;
  label: string;
  xAxesTitle: string;
  yAxesTitle: string;
  barChart: ReturnType<typeof UseStatesHook<ChartDataType>>;
};

const BarChart = ({
  chartData,
  title,
  label,
  xAxesTitle,
  yAxesTitle,
  barChart,
}: PieChartPropsType) => {
  const [loading, setLoading] = useState<boolean>(true);
  useEffect(() => {
    if (chartData) setLoading(false);
  }, [chartData]);
  const options: ChartOptions | undefined = {
    responsive: true,
    plugins: {
      legend: {
        position: "top",
      },
      title: {
        display: false,
        text: label,
      },
    },
    scales: {
      x: {
        title: {
          display: true,
          text: xAxesTitle,
          color: "#19504C",
        },
      },
      y: {
        title: {
          display: true,
          text: yAxesTitle,
          color: "#19504C",
        },
      },
    },
  };
  const data = {
    labels: chartData?.labels,
    datasets: [
      {
        label: label,
        data: chartData?.data,
        backgroundColor: "#96CCA3",
      },
    ],
  };
  return (
    <Card
      className="w-[49%] min-h-[30vh] max-desktop:w-[48%] max-tablet:basis-9/12 max-mobile:basis-[98%] overflow-hidden"
      bordered={false}
    >
      <ErrorBoundary
        error={barChart.error}
        refreshComponent={() => barChart.setRefresh((prev) => !prev)}
      >
        <CardTitle className="mb-14">{title}</CardTitle>
        {loading ? (
          <Skeleton.Avatar className="mt-2 ml-4" active shape="square" />
        ) : (
          <div className="overflow-scroll">
            <Bar options={options} data={data} width={"100%"} height={"60%"} />
          </div>
        )}
      </ErrorBoundary>
    </Card>
  );
};

export default BarChart;
