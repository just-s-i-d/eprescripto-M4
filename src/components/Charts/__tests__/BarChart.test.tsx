import { fireEvent, render, screen, waitFor } from "@testing-library/react";
import "@testing-library/jest-dom";
import BarChart from "../BarChart";
import useStatesHook from "@hooks/useStatesHook";
import { ApiResponseDataType, ChartDataType } from "@constants/types";

jest.mock("@hooks/useStatesHook");

describe("BarChart Component", () => {
  const chartData = {
    id: 1,
    attributes: {
      labels: ["Label1", "Label2", "Label3", "Label4", "Label5"],
      data: [10, 20, 30, 15, 5],
      totalCount: 80,
    },
  };
  it("Test to check if it renders correctly or not", () => {
    (useStatesHook as jest.Mock).mockResolvedValue({
      data: chartData,
      error: false,
      loading: true,
      setRefresh: jest.fn(),
    });
    const barChart = useStatesHook<ApiResponseDataType<ChartDataType>>();
    const { asFragment } = render(
      <BarChart
        chartData={chartData}
        title="Test Title"
        label="Test Label"
        xAxesTitle="X Axes Title"
        yAxesTitle="Y Axes Title"
        barChart={barChart}
      />,
    );
    expect(screen.queryByText("Something went wrong")).toBeNull();
    expect(asFragment()).toMatchSnapshot();
  });
  it("Test to check error handling", async () => {
    const mockSetRefresh = jest.fn();
    (useStatesHook as jest.Mock).mockReturnValue({
      data: {},
      error: true,
      setRefresh: mockSetRefresh,
    });
    const barChart = useStatesHook<ApiResponseDataType<ChartDataType>>();
    render(
      <BarChart
        chartData={chartData}
        title="Test Title"
        label="Test Label"
        xAxesTitle="X Axes Title"
        yAxesTitle="Y Axes Title"
        barChart={barChart}
      />,
    );
    await waitFor(() => {
      expect(screen.getByText("Something went wrong")).toBeInTheDocument();
      fireEvent.click(screen.getByRole("button", { name: "Try again" }));
      expect(mockSetRefresh).toHaveBeenCalled();
    });
  });
});
