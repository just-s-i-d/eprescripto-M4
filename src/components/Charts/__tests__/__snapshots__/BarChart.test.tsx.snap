// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BarChart Component Test to check if it renders correctly or not 1`] = `
<DocumentFragment>
  <div
    class="ant-card w-[49%] max-xl:w-[48%] max-xxl:w-[48%] max-md:w-9/12 max-sm:w-full css-dev-only-do-not-override-1adbn6x"
  >
    <div
      class="ant-card-body"
    >
      <h1
        class="text-xl font-bold mb-6 mb-14"
      >
        Test Title
      </h1>
      <canvas
        data-testid="bar-chart"
        height="150"
        role="img"
        width="300"
      />
    </div>
  </div>
</DocumentFragment>
`;
